### PR TITLE
Issue #2316 - Activatable Entity: ActivePropertyValidator to include DeactivatableDependencies for cascaded validation

### LINKED ISSUE #2316: Activatable Entity: ActivePropertyValidator to include DeactivatableDependencies for cascaded validation
### Description

`ActivePropertyValidator` does not process `@DeactivatableDependencies`, which may be defined for an activatable entity being deactivated, for determining "cascaded" active dependencies. This leads to situations where entity deactivation is prevented not as the result of validation of its property `active`, but as the result of cascaded processing of dependencies upon entity saving.

Determining and displaying a complete list of dependencies that prevent deactivation, which would include dependencies stemming from entities specified in `@DeactivatableDependencies`, would present a more complete picture of the situation as part of validation.

For example, deactivating a `Person` with the following definition may not have direct dependencies that prevent deactivation, but may have dependencies for person's roles as `InternalPerson`, `CustomerPerson`, or `ManagerPerson` that would prevent deactivation. 
```Java
@DeactivatableDependencies({ InternalPerson.class, CustomerPerson.class, ManagerPerson.class })
public class Person extends ActivatableAbstractEntity<String> {
...
```
The following needs to be implemented.

- [x] 1. Enhance `ActivePropertyValidator` to process entities from `@DeactivatableDependencies` for possible active dependencies that would prevent cascaded deactivation. Such entities may themselves have `@DeactivatableDependencies` and so this information needs to be processed recursively.

- [x] 2. The question of performance should be considered as such exhaustive search for active dependencies could require non-trivial computations, including execution of database queries. If performance testing would indicate issues, an approach where dependencies are computed and displayed for one entity at a time should be considered.

#### Implementation details

* `DomainEntitiesDependenciesUtils` is responsible for determining dependent entities and composing an EQL query that calculated the number of active dependencies. 
* `DomainEntityDependencies` captures and refines the dependency information, which explicitly excludes deactivatable dependencies from consideration as those would be deactivated automatically (this is corrected and expected). However, it includes a set of deactivatable dependencies as `automaticallyDeactivatedDependencies`, which can be used to extend the validation to those entities.

#### Change overview
The core change in this issue pertains to `ActivePropertyValidator`.
Test `transitive_deactivatable_dependencies_are_considered_when_validating_property_active` in `HandlingOfDeactivatableDependenciesTest` covers the case of transitive deactivatable dependencies.

### Expected outcome
A more complete list of active dependencies, preventing entity deactivation, displayed upon validation.

### PR DESCRIPTION
**Pull request by 01es**
Resolve #2316

# To be completed by the pull request creator

This section should be completed with reference to section [Preparing PR](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews#preparing-pr) of the [Code and PR reviews](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews) wiki page.

- [x] Converted PR to draft while it is being prepared by tapping the "Convert to draft" link beneath the "Reviewers" section.

- [x] A self-review of all changes has been completed, and the changes are in sync with the issue requirements.

- [x] Changes to the requirements have been reflected in the issue description.

- [x] Any "leftovers" such as sysouts, printing of stack traces, and any other "temporary" code, have been removed.

- [x] Minor refactorings, such as renamings, extraction of constants, etc., have been addressed.

- [x] Developer documentation (e.g., comments, Javadoc), have been provided where required.

- [x] New tests have been written, if required, to cover the new functionality.

- [x] All tests (all existing tests and all new tests) pass successfully.

- [x] Changes have been inspected for possible NPE situations, and the changes are sufficiently defensive.

- [x] The correct base branch has been selected for these changes to be merged into.

- [x] The latest changes from the base branch have already been merged into this feature branch (and tested).

- [x] Added a change overview to the issue description or as a wiki page, referenced in the issue description.
      Some issue might be very descriptive and server in place of a wiki page.
      In such cases consider adding label `Wiki like` to the issue.

- [x] Changes subject to performance considerations have been evaluated, and tested against production-size data if applicable.

- [x] This pull request does not contain significant changes, and at least one appropriate reviewer has been selected.

- [x] The `In progress` label has been removed from the issue.

- [x] The `Pull request` label has been added to the issue.

- [x] Made PR ready for review by tapping the "Ready for review" button below the list of commits on the PR page.

# To be completed by the pull request reviewer

This section should be completed with reference to section [Performing PR review](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews#performing-pr-review) of the [Code and PR reviews](https://github.com/fieldenms/devops/wiki/Code-and-PR-reviews) wiki page.

- [x] The `In progress` label has been added to the pull request in GitHub.

- [x] The issue requirements have been read and understood (along with any relevant emails and/or Slack messages).

- [x] The correct base branch is specified, and that base branch is up-to-date in the local source.

- [x] The issue branch has been checked out locally, and had the base branch merged into it.

- [x] All automated tests pass successfully.

- [x] Ensure the implementation satisfies the functional requirements.

- [x] Ensure that code changes are secure and align with the established coding practices, including code formatting and naming conventions.

- [x] Ensure that code changes are documented and covered with automated tests as applicable.

- [x] Ensure that code changes are well-suited for informal reasoning.

- [x] Ensure that changes are documented for the end-user (a software engineer in the case of TG, or an application user in the case of TG-based applications).

- [x] If there are significant changes (described above), special attention has been paid to them.
      Marked the task items in section "Significant changes" as completed to indicated that corresponding changes have been reviewed, improved if necessary, and approved.

- [x] The issue or issues addressed by the pull request are associated with the relevant release milestone.

# To be completed by the pull request reviewer once the changes have been reviewed and accepted

- [ ] The changes have been merged into the base branch (unless there is a specific request not to do so, e.g., they are to be released to SIT).

- [ ] The issue branch has been deleted (unless the changes have not been merged - see above, or there is a specific request not to do so).

- [ ] The `In progress` label has been removed from the pull request.

- [ ] The `Pull request` label has been removed from the pull request.

### CODE REVIEW

File: platform-pojo-bl/src/main/java/ua/com/fielden/platform/utils/Validators.java
Comment (elvin-fms) [inline platform-pojo-bl/src/main/java/ua/com/fielden/platform/utils/Validators.java:1]: So many changes in here, but many of them look like just formatting tweaks, rephrasing of comments and similar.
Therefore it is challenging for me to assess the impact of these changes.

**Review by elvin-fms — CHANGES_REQUESTED** [review]
I still don't understand how the `refCount` property is used now - in `ActivePropertyValidator` all code references to it appear to have been removed, it is only mentioned in comments and messages.

File: platform-pojo-bl/src/main/java/ua/com/fielden/platform/utils/Validators.java
Comment (01es) [reply platform-pojo-bl/src/main/java/ua/com/fielden/platform/utils/Validators.java:1]: @elvin-fms those utility method have reasonable test coverage. Of course, you could always ascertain whether the original semantics hold after the refactoring by comparing the code before the changes and in their current form.

File: platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/validation/ActivePropertyValidator.java [hunk 1]
```diff
@@ -26,130 +38,106 @@
...
-        if (!newValue && entity.isPersisted()) { // a persisted entity is being deactivated, but it may still be referenced
-            // let's check refCount... it could potentially be stale...
-            final IEntityDao<?> co = coFinder.find(entity.getType());
-            final int count;
-            if (!co.isStale(entity.getId(), entity.getVersion())) {
-                count = entity.getRefCount();
+        // A persisted entity is being deactivated, but it may still be referenced.
+        if (!newValue && entity.isPersisted()) {
+            // Check refCount... it could potentially be stale...
+            final IEntityReader<?> co = co(entity.getType());
```
[the remark this answers is not in the document]
Comment (01es) [reply platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/validation/ActivePropertyValidator.java:66]: `ActivePropertyValidator` indeed now computes references dynamically. The comment was missed and shall be removed.

There were no changes to `refCount` computations. 
 If you would like to learn how is it still used, please refer to various methods in `PersistentEntitySaver`, such as `collectActivatableDirtyProperties`, `shouldProcessAsActivatable`, `handleDirtyActivatableProperty`, and `handleNonDirtyActivatableIfNecessary`.

Potential removal of `refCount` is covered by issue https://github.com/fieldenms/tg/issues/1745

[the remark this answers is not in the document]
Comment (01es) [reply platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/validation/ActivePropertyValidator.java]: The comment was removed https://github.com/fieldenms/tg/commit/5f50959a61a985c76a22009fa8efa3780418d878

[the remark this answers is not in the document]
Comment (01es) [reply platform-pojo-bl/src/main/java/ua/com/fielden/platform/entity/validation/ActivePropertyValidator.java]: Also https://github.com/fieldenms/tg/commit/43231577b6ea15c60b43fc1ad2c475f981eb3fc7

File: platform-pojo-bl/src/main/java/ua/com/fielden/platform/utils/Validators.java
Comment (elvin-fms) [reply platform-pojo-bl/src/main/java/ua/com/fielden/platform/utils/Validators.java:1]: Yes, that's exactly what I tried.

**Review by elvin-fms — APPROVED** [review]

**Pull request merged**